include/cifer/abe/dippe.h

```diff
@@ -0,0 +1,276 @@
+*/
+
+/**
+ * cfe_dippe represents the DIPPE scheme
+ */
+typedef struct cfe_dippe {
+    size_t assump_size;
+    mpz_t p;
+    cfe_mat_G1 g1_A;
+    cfe_mat_G1 g1_UA;
+} cfe_dippe;
+
+/**
+ * Initializes the DIPPE scheme and generates the global parameters
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param assump_size The size of the underlying assumption
+ */
+void cfe_dippe_init(cfe_dippe *dippe, size_t assump_size);
+
+/**
+ * Clears the scheme and frees allocated memory
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ */
+void cfe_dippe_free(cfe_dippe *dippe);
+
+/**
+ * cfe_dippe_attribute_vector represents an attribute vector that is used
+ * to request attributes from an authority
+ */
+typedef struct cfe_dippe_attribute_vector {
+    char *s;
+    size_t len;
+    mpz_t attrs;
+} cfe_dippe_attribute_vector;
+
+/**
+ * Prepares the attribute vector struct
+ *
+ * @param av A pointer to a cfe_dippe_attribute_vector struct; Represents the resulting attribute vector
+ * @param vec_len Length of the resulting attribute vector
+ * @param pattern Array of indices to define the attribute vector
+ * @param pat_len Length of the pattern array
+ * @return Error code
+ */
+cfe_error cfe_dippe_attribute_vector_init(cfe_dippe_attribute_vector *av, size_t vec_len, size_t pattern[], size_t pat_len);
+
+/**
+ * Clears the attribute vector and frees allocated memory
+ *
+ * @param av A pointer to a cfe_dippe_attribute_vector struct
+ */
+void cfe_dippe_attribute_vector_free(cfe_dippe_attribute_vector *av);
+
+/**
+ * cfe_dippe_policy_vector represents a policy vector that is used
+ * during encryption
+ */
+typedef struct cfe_dippe_policy_vector {
+    size_t len;
+    cfe_vec pol;
+} cfe_dippe_policy_vector;
+
+/**
+ * Prepares the policy vector struct to used as threshold policy
+ *
+ * @param pv A pointer to a cfe_dippe_policy_vector struct; Represents the resulting policy vector
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param vec_len Length of the resulting policy vector
+ * @param pattern Array of indices to define the policy vector
+ * @param pat_len Length of the pattern array
+ * @param threshold Theshold value
+ * @return Error code
+ */
+cfe_error cfe_dippe_threshold_policy_vector_init(cfe_dippe_policy_vector *pv, cfe_dippe *dippe, size_t vec_len, size_t pattern[], size_t pat_len, size_t threshold);
+
+/**
+ * Prepares the policy vector struct to used as conjunction policy
+ *
+ * @param pv A pointer to a cfe_dippe_policy_vector struct; Represents the resulting policy vector
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param vec_len Length of the resulting policy vector
+ * @param pattern Array of indices to define the policy vector
+ * @param pat_len Length of the pattern array
+ * @return Error code
+ */
+cfe_error cfe_dippe_conjunction_policy_vector_init(cfe_dippe_policy_vector *pv, cfe_dippe *dippe, size_t vec_len, size_t pattern[], size_t pat_len);
```

Comment:
Maybe rename parameter `vec_len` to something more descriptive, like `num_attrib` for number of all attributes. Then `vec_len = num_attrib + 1`. Similarly with threshold.

Reply:
Done.